roofit/roofitcore/src/RooDataHist.cxx

```diff
@@ -987,8 +987,10 @@ Int_t RooDataHist::getIndex(const RooAbsCollection& coord, Bool_t fast) const {
 /// as the internal coordinates. In this case, values are looked up only by index.
 std::size_t RooDataHist::calcTreeIndex(const RooAbsCollection& coords, bool fast) const
 {
-  // With fast, caller promises that layout of "coords" is identical to our internal "vars"
-  assert(!fast || _vars.size() == coords.size());
+  // Asserts commented out because the code is performance critical.
+
+  // With fast, caller promises that layout of "coords" is identical to our internal "vars".
+  // assert(!fast || _vars.size() == coords.size());
```

Comment:
Check this out: [assert](https://en.cppreference.com/w/cpp/error/assert)
It's the magic weapon for things that "should", but you cannot be sure that they "are", unless you get c++ contracts.
Note that there is **no** run time overhead in release builds - they don't generate *any* code.

Now, imagine the following:
A user comes to the forum and tells you that `xy` yields wrong results when they do A, but with B it works. It doesn't crash, the results are just wrong, and nobody knows where the error happens.
Those are the bugs that take long to track down.
You ask for the code, run it *in your debug build that you should always have around*, the assert fails  and you tell the user what's wrong or fix the bug.

Long story short:
The asserts should stay in, and please use them to assert all things where you go "hmm, this *should* be OK if people (or you) used it correctly".
For things that are not performance critical, it's of course even better to validate inputs before you start a fit etc.

Reply:
> For things that are not performance critical, it's of course even better to validate inputs before you start a fit etc.

I mean run-time checks that are active also in release builds, but this is mostly for the initialisation phase, while the asserts are better in the execution phase when you want speed.